subprojects/maven/src/main/java/org/gradle/api/publish/maven/internal/validation/MavenPublicationErrorChecker.java

```diff
@@ -43,14 +48,56 @@ public abstract class MavenPublicationErrorChecker extends PublicationErrorCheck
      * @throws PublishException if the artifacts are modified
      */
     public static void checkThatArtifactIsPublishedUnmodified(PublishArtifact source, DefaultMavenArtifactSet mainArtifacts) {
+        // Do we want to allow _additional_ artifacts to published? If not, we need to completely associate the source and artifacts
+        // at a higher level. As-is, this just verifies that none are removed from the components artifacts.
+        MavenArtifact bestMatch = null;
+        Set<Difference> bestMatchDifferences = null;
         for (MavenArtifact mavenArtifact : mainArtifacts) {
-            if (source.getFile().equals(mavenArtifact.getFile())
-                && source.getExtension().equals(mavenArtifact.getExtension())
-                && Strings.nullToEmpty(source.getClassifier()).equals(Strings.nullToEmpty(mavenArtifact.getClassifier()))) {
+            EnumSet<Difference> difference = EnumSet.noneOf(Difference.class);
+            if (!source.getFile().equals(mavenArtifact.getFile())) {
+                difference.add(Difference.FILE);
+            }
+            if (!source.getExtension().equals(mavenArtifact.getExtension())) {
```

Comment:
🤔 Is this even a necessary sub-case?  Why does this need to be distinct from the  File differs case?

The original code compared extensions separately, but it seems to me like that was redundant and could be dropped as it is implicitly part of the File check.

Reply:
It can actually differ, I have changed the extension test case to show how.

Reply:
Ah, ok: `artifact source: "customFile.txt", extension: "csv"`, I didn't realize you could lie to the publication about the extension like that.